### PR TITLE
openstack: 4.12+ jobs to use application credentials

### PR DESCRIPTION
**Pull request by pierreprinetti**
This change introduces a new environment variable for OpenStack jobs: `OPENSTACK_AUTHENTICATION_METHOD`, which defaults to `application-credentials`.

Password authentication is enforced if the detected OpenShift version is below 4.12.

[OSASINFRA-3001](https://issues.redhat.com//browse/OSASINFRA-3001)

---

In a separate commit, silence the MAPO failure to pick up new credentials: see
* [the bug report](https://issues.redhat.com/browse/OCPBUGS-8687)
* [the open task](https://issues.redhat.com/browse/OSASINFRA-3118) to revert the workaround

### CODE REVIEW

**Comment by pierreprinetti** [comment]
/hold

This change requires all base images to include `yq`, and it's not always the case.

**Comment by pierreprinetti** [comment]
`yq` issue solved for shiftstack-ci.
Now on to the next: the `openstack-conf-clouds` step does not use `ca-cert` to connect to openstack.

**Comment by pierreprinetti** [comment]
/hold cancel
Let's see what the CI tells us

**Comment by pierreprinetti** [comment]
/retest
No job < 4.12 has succeeded yet.

File: ci-operator/config/shiftstack/shiftstack-ci/shiftstack-shiftstack-ci-main.yaml [hunk 1]
```diff
@@ -12,6 +12,7 @@ images:
 - dockerfile_literal: |-
     FROM src
     RUN ./hack/dnf_safe -y install python3-openstackclient python3-octaviaclient jq && dnf clean all
+    RUN python -m pip install yq
```
Comment (mandre) [inline ci-operator/config/shiftstack/shiftstack-ci/shiftstack-shiftstack-ci-main.yaml:15]: It looks like you changed your mind and used `jq` in the end. There's no need to install `yq` in that case.

File: ci-operator/config/shiftstack/shiftstack-ci/shiftstack-shiftstack-ci-main__periodic-4.12.yaml [hunk 2]
```diff
@@ -141,12 +141,12 @@ tests:
...
     workflow: openshift-e2e-openstack-upi
-- as: e2e-openstack-appcreds
+- as: e2e-openstack-password
   minimum_interval: 24h
   steps:
     cluster_profile: openstack-vexxhost
     env:
-      APPLICATION_CREDENTIALS: "true"
+      PASSWORD_AUTHENTICATION: "true"
       TEST_SKIPS: 'In-tree Volumes \[Driver: cinder\]'
```
Comment (mandre) [inline ci-operator/config/shiftstack/shiftstack-ci/shiftstack-shiftstack-ci-main__periodic-4.12.yaml:150]: When switching our jobs to use app creds by default, we need to skip the in-tree cinder tests (as we were doing here) as the cinder executable does not work with app creds. Possibly something we'll want [at a higher level in origin](https://github.com/openshift/origin/blob/master/test/extended/util/annotate/rules.go).

File: ci-operator/step-registry/openstack/conf/clouds/openstack-conf-clouds-commands.sh [hunk 3]
```diff
@@ -7,14 +7,27 @@ set -o pipefail
...
+
 cp "/var/run/cluster-secrets/${CLUSTER_TYPE}/clouds.yaml" "$OS_CLIENT_CONFIG_FILE"
 
 if [ -f "/var/run/cluster-secrets/${CLUSTER_TYPE}/osp-ca.crt" ]; then
 	cp "/var/run/cluster-secrets/${CLUSTER_TYPE}/osp-ca.crt" "${SHARED_DIR}/osp-ca.crt"
 	sed -i "s|cacert: .*|cacert: ${SHARED_DIR}/osp-ca.crt|" "$OS_CLIENT_CONFIG_FILE"
 fi
 
-if [[ -n $APPLICATION_CREDENTIALS ]]; then
+if ! is_openshift_version_gte "4.12"; then
```
Comment (mandre) [inline ci-operator/step-registry/openstack/conf/clouds/openstack-conf-clouds-commands.sh:25]: Could you add a comment explaining the reasons we have this check? Could be as simple as saying that application credentials is supported starting from 4.12.

File: ci-operator/config/shiftstack/shiftstack-ci/shiftstack-shiftstack-ci-main.yaml (hunk 1 above)
Comment (mandre) [reply ci-operator/config/shiftstack/shiftstack-ci/shiftstack-shiftstack-ci-main.yaml:15]: Gotcha!

**Comment by pierreprinetti** [comment]
/pj-rehearse periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-password periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-serial periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.11-e2e-openstack-serial pull-ci-openshift-cluster-api-provider-openstack-release-4.13-e2e-openstack-proxy pull-ci-openshift-cluster-api-provider-openstack-release-4.10-e2e-openstack-proxy pull-ci-openshift-cluster-api-provider-openstack-release-4.12-e2e-upgrade

**Comment by pierreprinetti** [comment]
Tests are successful. I am force-pushing
* better comments
* if method is set to password and the orignal clouds.yaml contains appcreds, then fail.

**Comment by pierreprinetti** [comment]
/hold cancel
/pj-rehearse periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-password periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-serial periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.11-e2e-openstack-ovn-serial

**Comment by pierreprinetti** [comment]
/pj-rehearse periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-password periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-serial periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.11-e2e-openstack-serial

**Comment by pierreprinetti** [comment]
Tests look OK. @mandre I think it’s ready to merge.

**Comment by pierreprinetti** [comment]
/pj-rehearse periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-serial

**Comment by pierreprinetti** [comment]
/pj-rehearse periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-serial

**Comment by pierreprinetti** [comment]
/hold cancel
/pj-rehearse periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-password periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-serial periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.11-e2e-openstack-serial

**Comment by pierreprinetti** [comment]
/pj-rehearse reject
(wait for CI results)

**Comment by pierreprinetti** [comment]
/pj-rehearse periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.11-e2e-openstack-serial

**Comment by pierreprinetti** [comment]
/pj-rehearse periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-password

**Comment by pierreprinetti** [comment]
/pj-rehearse periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-serial

**Comment by pierreprinetti** [comment]
/pj-rehearse periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-password periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-serial periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.11-e2e-openstack-serial

**Comment by pierreprinetti** [comment]
/pj-rehearse periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-password periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-serial

**Comment by pierreprinetti** [comment]
/pj-rehearse periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-password periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.13-e2e-openstack-ovn-serial

File: ci-operator/step-registry/openstack/conf/clouds/openstack-conf-clouds-commands.sh [hunk 4]
```diff
@@ -45,34 +57,40 @@ new_clouds_yaml() {
...
+			echo 'Environment variable ROTATE_CLOUD_CREDENTIALS detected. Generating a set of application credentials for the rotation.'
+			new_application_credentials "$clouds_yaml" > "${SHARED_DIR}/clouds2.yaml"
+			new_application_credentials "$clouds_yaml" --unrestricted > "${SHARED_DIR}/clouds-unrestricted.yaml"
 
-	declare appcreds_id
-	appcreds_id="$(yq -r ".clouds.\"${OS_CLOUD}\".auth.application_credential_id" "${SHARED_DIR}/clouds2.yaml")"
-	echo "Generated additional application credentials with ID ${appcreds_id}"
-fi
+			declare appcreds_id
+			appcreds_id="$(yq -r ".clouds.\"${OS_CLOUD}\".auth.application_credential_id" "${SHARED_DIR}/clouds2.yaml")"
```
Comment (mandre) [inline ci-operator/step-registry/openstack/conf/clouds/openstack-conf-clouds-commands.sh:80]: nit, we're shadowing `appcreds_id` but I don't think it matters.

**Review by mandre — COMMENTED** [review]
/lgtm

**Pull request merged**